Form linking UI and backend

benrudolph opened the pull request:
@snopoke here's the UI and saving to the backend. I'd recommend taking a look at the first commit rather than the whole thing since adding the feature flag obfuscates a few parts.

Here are some screenshots of the UI:
![screen shot 2015-03-02 at 2 37 44 pm](https://cloud.githubusercontent.com/assets/918514/6448753/28433096-c0ea-11e4-8c8c-ddefbd13dc4c.png)
With errors:
![screen shot 2015-03-02 at 2 37 37 pm](https://cloud.githubusercontent.com/assets/918514/6448758/2f4e00a0-c0ea-11e4-8736-948453e68d5e.png)

Open questions:
- What's the best way to do the syntax checking for the XPath expression? (I think we said we'd ensure that it's proper syntax, but not ensure that it's a valid XPath expression that will evaluate to something)
- Is there an example you could point me to of how we do JS translation? Or is it mainly ad-hock 
- `linkable_forms` seems like it could be an expensive operation and could potentially add a lot of data to the response object. I couldn't seem to figure out a way to use the existing response in the django template to generate the `linkable_forms`. Let me know if you know of a better way.
- Do we have a way to add JS unit tests? I mainly want to put some in for the ViewModel. `django-qunit` looks like a good candidate, but this is probably a larger discussion

benrudolph committed:
Functionality for adding and removing Form Links

benrudolph committed:
Remove extra space

benrudolph committed:
Put behind FORM_LINK_WORKFLOW toggle

dimagimon commented:
There are 22 errors:

* corehq/apps/app_manager/views.py, line 1733 - E201 whitespace after '{'
* corehq/apps/app_manager/views.py, line 1733 - E202 whitespace before '}'
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 1 - Use the function form of "use strict".
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 19 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 24 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 36 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 59 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 74 - Missing semicolon.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 9 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 14 - '$' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 17 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 18 - '_' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 21 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 21 - '_' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 44 - '_' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 53 - '$' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 70 - '_' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 77 - 'window' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 84 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 85 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 88 - 'ko' is not defined.
* corehq/apps/app_manager/static/app_manager/js/form_workflow.js, line 92 - '_' is not defined.

benrudolph committed:
Fix dimagimon errors

benrudolph committed:
Wrap FormWorkflow in anonymous function

This allows us to use ‘use strict’ just once

benrudolph committed:
Remove extraneous comment

snopoke commented on corehq/apps/app_manager/static/app_manager/js/form_workflow.js:
```diff
@@ -0,0 +1,107 @@
+    };
+
+    FormWorkflow.Form = function(form, options) {
+        this.name = window.localize(form.name, options.locale);
```
Where does this `localize` function get defined?

benrudolph replied:
https://github.com/benrudolph/commcare-hq/blob/form-link-158797/corehq/apps/cloudcare/static/cloudcare/js/util.js#L7

I found it when searching for ways we were localizing. Probably not the best to use, I don't think it's loaded on every page and it's not clear that it even should be loaded on the page. Will change

snopoke replied:
yea, I'm pretty sure that isn't loaded on this page (or at least it shouldn't be)

snopoke replied:
Since the user can't change the language without reloading the page you could just to it in python. e.g. https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/views.py#L818-818

benrudolph replied:
ah good point, i'll use that

snopoke commented on corehq/apps/app_manager/views.py:
```diff
@@ -961,6 +962,13 @@ def view_generic(req, domain, app_id=None, module_id=None, form_id=None, is_user
+        if toggles.FORM_LINK_WORKFLOW.enabled(req.user.username):
+            modules = filter(lambda m: m.case_type == module.case_type, app.get_modules())
+            context.update({
+                'linkable_forms': sum([list(m.get_forms()) for m in modules], [])
```
Can use `itertools.chain`:
```python
list(itertools.chain(list(m.get_forms()) for m in modules))
```

benrudolph replied:
cool this is definitely more readable

snopoke commented on corehq/apps/app_manager/views.py:
```diff
@@ -961,6 +962,13 @@ def view_generic(req, domain, app_id=None, module_id=None, form_id=None, is_user
         })
+
+        if toggles.FORM_LINK_WORKFLOW.enabled(req.user.username):
+            modules = filter(lambda m: m.case_type == module.case_type, app.get_modules())
```
This won't work for `AdvancedForms` but I'm not 100% clear on what the inclusion criteria are so this will probably do for now.

benrudolph replied:
I'd be interested to sync when you've nailed down the inclusion criteria so I have a better understanding of the bigger picture

snopoke commented on corehq/apps/app_manager/templates/app_manager/partials/form_workflow.html:
```diff
@@ -11,12 +11,82 @@
+                value: formLink.formId
+                "></select>
+            <button class="btn btn-danger" data-bind="
+                event: { click: $parent.onDestroyFormLink.bind($parent) }
```
You can just use the `click` binding: http://knockoutjs.com/documentation/click-binding.html

benrudolph replied:
yep good catch

snopoke commented on corehq/apps/app_manager/templates/app_manager/partials/form_workflow.html:
```diff
@@ -11,12 +11,82 @@
+    </script>
+    <script type="text/html" id="form-link-error-template">
+        <li data-bind="text: error"></li>
+    </script>
```
Since this is just one line we probably don't need a template here and can just do a `foreach`

snopoke commented on corehq/apps/app_manager/static/app_manager/js/form_workflow.js:
```diff
@@ -0,0 +1,107 @@
+    };
+
+
+    FormWorkflow.prototype.workflowDisplay = function() {
```
This could be a `computed observable`: http://knockoutjs.com/documentation/computedObservables.html

snopoke commented on corehq/apps/app_manager/static/app_manager/js/form_workflow.js:
```diff
@@ -0,0 +1,107 @@
+        });
+    };
+
+    FormWorkflow.prototype.onWorkflowChange = function(workflow, event) {
```
You could just subscribe to the workflow observable:
```javascript
self.workflow.subscribe(function(value) {
    this.showFormLinkUI(value === FormWorkflow.Values.FORM);
});
```

snopoke commented:
Generally looks good. I like the extraction of the JS to a separate file. None of my comments are blockers except perhaps the one re `window.localize`.

Re your questions:
- What's the best way to do the syntax checking for the XPath expression? (I think we said we'd ensure that it's proper syntax, but not ensure that it's a valid XPath expression that will evaluate to something)
  - I had thought we were doing something like that but I can't find it. One possibility is to just test the expression on a bogus doc which will generate an error if the expression is invalid: http://lxml.de/xpathxslt.html#xpath
- Is there an example you could point me to of how we do JS translation? Or is it mainly ad-hock 
  - We don't do this at all. Django does have a way to do it but we've never implemented it (despite it being on our tech dept roadmap: https://docs.djangoproject.com/en/1.7/topics/i18n/translation/#internationalization-in-javascript-code)
- `linkable_forms` seems like it could be an expensive operation and could potentially add a lot of data to the response object. I couldn't seem to figure out a way to use the existing response in the django template to generate the `linkable_forms`. Let me know if you know of a better way.
  - You don't need to send the entire form, just the `name` and `id`
- Do we have a way to add JS unit tests? I mainly want to put some in for the ViewModel. `django-qunit` looks like a good candidate, but this is probably a larger discussion
  - I think this is also on the tech debt roadmap and it would be great to see but I think its outside the scope of this PR. @millerdev may have some thoughts on the topic.

benrudolph committed:
Code review changes

- Form link, hard dependency on list of forms
- subscribe to workflow variable
- workflowDisplay as computed
- remove localization in JS
- use click event binding
- foreach instead of template
- better linkable_forms implementation

benrudolph commented:
For now should be OK, unless I missed something from the edits you suggested. I need to look into XPath validation, but I think that could come in a separate PR if you'd like to use this to do the suite-xml portion

snopoke commented:
Looks good. Not 100% sure what the QA path for this is. Perhaps just check with Amelia / Christy if they want to take a pass on staging.

benrudolph commented:
Will speak to Christy tomorrow about staging

benrudolph commented:
But also we may want to wait until the feature is fully implemented to save them from going through it twice?

snopoke commented:
Merging this now. Can QA on prod when the rest of the feature is done.
